ethcore/types/src/transaction/transaction.rs

```diff
@@ -259,35 +433,204 @@ impl Transaction {
+            Self::Legacy(tx) => tx,
+            Self::AccessList(ocl) => ocl.tx(),
+        }
+    }
+
+    pub fn tx_mut(&mut self) -> &mut Transaction {
+        match self {
+            Self::Legacy(tx) => tx,
+            Self::AccessList(ocl) => ocl.tx_mut(),
+        }
+    }
+
+    fn decode_new(tx: &[u8]) -> Result<UnverifiedTransaction, DecoderError> {
+        if tx.is_empty() {
+            // at least one byte needs to be present
+            return Err(DecoderError::RlpIncorrectListLen);
+        }
+        let id = tx[0].try_into();
+        if id.is_err() {
+            return Err(DecoderError::Custom("Unknown transaction"));
+        }
+        //other transaction types
+        match id.unwrap() {
+            TypedTxId::AccessList => AccessListTx::decode(&tx[1..]),
+            TypedTxId::Legacy => return Err(DecoderError::Custom("Unknown transaction")),
+        }
+    }
+
+    pub fn decode(tx: &[u8]) -> Result<UnverifiedTransaction, DecoderError> {
+        if tx.is_empty() {
+            // at least one byte needs to be present
+            return Err(DecoderError::RlpIncorrectListLen);
+        }
+        let header = tx[0];
+        //type of transaction can be obtained from first byte. If first bit is 1 it means we are dealing with RLP list.
+        //if it is 0 it means that we are dealing with custom transaction defined in EIP-2718.
+        if (header & 0x80) != 0x00 {
+            Transaction::decode(&Rlp::new(tx))
+        } else {
+            Self::decode_new(tx)
+        }
+    }
+
+    pub fn decode_rlp(tx: &Rlp) -> Result<UnverifiedTransaction, DecoderError> {
+        if tx.is_list() {
+            //legacy transaction wrapped around RLP encoding
+            Transaction::decode(tx)
+        } else {
+            Self::decode_new(tx.data()?)
+        }
+    }
+
+    fn rlp_append(
+        &self,
+        s: &mut RlpStream,
+        chain_id: Option<u64>,
+        signature: &SignatureComponents,
+    ) {
+        match self {
+            Self::Legacy(tx) => tx.rlp_append(s, chain_id, signature),
+            Self::AccessList(opt) => opt.rlp_append(s, chain_id, signature),
+        }
+    }
 }
 
-/// Signed transaction information without verified signature.
+impl HeapSizeOf for TypedTransaction {
+    fn heap_size_of_children(&self) -> usize {
+        match self {
+            TypedTransaction::Legacy(legacy) => legacy.heap_size_of_children(),
+            TypedTransaction::AccessList(oal) => oal.tx().heap_size_of_children(),
+        }
+    }
+}
+
+/// Components that constitute transaction signature
 #[derive(Debug, Clone, Eq, PartialEq)]
-pub struct UnverifiedTransaction {
-    /// Plain Transaction.
-    unsigned: Transaction,
+pub struct SignatureComponents {
     /// The V field of the signature; the LS bit described which half of the curve our point falls
     /// in. The MS bits describe which chain this transaction is for. If 27/28, its for all chains.
-    v: u64,
+    v: u8,
```

Comment:
to avoid futher confustions, I would prefer to rename `v` to `stadard_v` or alike

Reply:
I will use `standard_v`